File: 981.java
import java.util.*;
import javafx.util.Pair;

class TimeMap {
    private Map<String, List<Pair<String, Integer>>> map;

    public TimeMap() {
        this.map = new HashMap<>();
    }

    public void set(String key, String value, int timestamp) {
        List<Pair<String, Integer>> result = this.map.getOrDefault(key, new ArrayList<>());
        result.add(new Pair<String, Integer>(value, timestamp));
        this.map.put(key, result);
    }

    public String get(String key, int timestamp) {
        List<Pair<String, Integer>> search = this.map.getOrDefault(key, new ArrayList<>());
        if (search.isEmpty() || search.get(0).getValue() > timestamp) {
            return "";
        }
        int left = 0, right = search.size() - 1;
        String result = "";
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (search.get(mid).getValue() <= timestamp) {
                result = search.get(mid).getKey();
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return result;
    }
}

File: Python/981.py
class TimeMap:

    def __init__(self):
        self.map = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.map:
            self.map[key] = []
        self.map[key].append([value, timestamp])

    def get(self, key: str, timestamp: int) -> str:
        result = ""
        values = self.map.get(key, [])
        left, right = 0, len(values) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if values[mid][1] <= timestamp:
                result = values[mid][0]
                left = mid + 1
            else:
                right = mid - 1
        return result


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
